laghos.cpp

```diff
@@ -784,6 +784,91 @@ int main(int argc, char *argv[])
+        if (romOptions.parameterID <= 0)
+        {
+            if (myid == 0)
+            {
+                std::ofstream outfile_offlineParam(offlineParam_outputPath);
+                outfile_offlineParam << romOptions.useOffset << " " << romOptions.offsetType << " " << romOptions.RHSbasis << " " << numWindows << " " << twfile << endl;
+                outfile_offlineParam << romOptions.parameterID << " " <<  rhoFactor << " " << blast_energyFactor << endl;
+                outfile_offlineParam.close();
+            }
+        }
+        else
+        {
+            std::ifstream infile_offlineParam(offlineParam_outputPath);
+            MFEM_VERIFY(infile_offlineParam.is_open(), "Offline parameter record file does not exist.");
+            std::string line;
+            std::vector<std::string> words;
+            std::getline(infile_offlineParam, line);
+            split_line(line, words);
+            MFEM_VERIFY(std::stoi(words[0]) == romOptions.useOffset, "-romos option does not match record.");
+            MFEM_VERIFY(std::stoi(words[1]) == romOptions.offsetType, "-romostype option does not match record.");
+            MFEM_VERIFY(std::stoi(words[2]) == romOptions.RHSbasis, "-romsrhs option does not match record.");
+            MFEM_VERIFY(std::stoi(words[3]) == numWindows, "-nwin option does not match record.");
+            MFEM_VERIFY(std::strcmp(words[4].c_str(), twfile) == 0, "-tw option does not match record.");
+            infile_offlineParam.close();
+            if (myid == 0)
+            {
+                std::ofstream outfile_offlineParam(offlineParam_outputPath, std::fstream::app);
+                outfile_offlineParam << romOptions.parameterID << " " <<  rhoFactor << " " << blast_energyFactor << endl;
+                outfile_offlineParam.close();
+            }
+        }
+    }
+
+    if (rom_online)
+    {
+        std::ifstream infile_offlineParam(offlineParam_outputPath);
+        MFEM_VERIFY(infile_offlineParam.is_open(), "Offline parameter record file does not exist.");
+        std::string line;
+        std::vector<std::string> words;
+        std::getline(infile_offlineParam, line);
+        split_line(line, words);
+        MFEM_VERIFY(std::stoi(words[0]) == romOptions.useOffset, "-romos option does not match record.");
+        MFEM_VERIFY(std::stoi(words[1]) == romOptions.offsetType, "-romostype option does not match record.");
+
+        if (romOptions.offsetType == interpolateOffset)
+        {
+            int true_idx = -1;
+            double coeff_sum = 0.0;
+            while (std::getline(infile_offlineParam, line))
+            {
+                split_line(line, words);
+                romOptions.paramID_list.push_back(std::stoi(words[0]));
+                double coeff = 0.0;
+                coeff += (rhoFactor - atof(words[1].c_str())) * (rhoFactor - atof(words[1].c_str()));
```

Comment:
What is the purpose of these calculations involving `rhoFactor` and `blast_energyFactor`?

Reply:
They are to compute the distances from online parameters to each offline parameter, which is further used to determine the coefficients with respect to the offline parameters in the interpolation of the offset in the online stage. The calculations may have to be relocated to `laghos_rom.cpp` for a better format.

Reply:
@siuwuncheung Thanks for the explanation. Should these distances by normalized by the range of the parameter space? For example, if `rhoFactor` ranges over (0.9, 1.1), should there be a normalization based on the range width 0.2?

Reply:
@dylan-copeland I have chosen the [inverse distance weighting](https://en.wikipedia.org/wiki/Inverse_distance_weighting) for interpolation. The advantages are that 
1. we can make sure the interpolation gives a convex combination of the data points, i.e. stored offsets, 
2. the normalization is provided automatically by sum of inverse distances, without the need of explicitly specifying the range of the varying parameter,
3. can be used for interpolation on multidimensional inputs, i.e. varying both blast energy and density simultaneously.

It is only a particular choice and may have its drawback though, so feel free to suggest another method if we want to compare. Also, may I ask if you think it is better to leave it the way as it is (in `laghos.cpp`), or to pass `rhoFactor` and `blast_energyFactor` to `ROMBasis` and compute the coefficients in each declaration of new basis (in `laghos_rom.cpp`)?

Reply:
@siuwuncheung Thanks, please just add a comment about the use of inverse distance weighting for interpolation. I agree that this piece of code should be in `laghos_rom.cpp`. It hurts the readability of the main function. Also, it seems the only inputs are the scalar factors, and the output is `romOptions`. This seems straightforward to put in the `ROM_Basis` constructor.

Reply:
@dylan-copeland Thanks for your suggestion. I added `rhoFactor` and `blast_energyFactor` to `romOptions`, made `coeff_List` and `paramID_List` private members of `ROMBasis` and relocated the calculation to `ROMBasis`. Hope it makes more sense now.